features/flows/classification/src/main/java/org/opennms/netmgt/flows/classification/DefaultClassificationEngine.java

```diff
@@ -0,0 +1,117 @@
+
+public class DefaultClassificationEngine implements ClassificationEngine {
+
+    private final ClassificationRuleDAO ruleDAO;
+
+    private final List<Classifier> classifierList = new ArrayList<>();
+
+    // This can be a DAO or similar in the future
+    public DefaultClassificationEngine(ClassificationRuleDAO ruleDAO) {
+        this.ruleDAO = Objects.requireNonNull(ruleDAO);
+        this.reload();
+    }
+
+    protected DefaultClassificationEngine(Supplier<List<Rule>> supplier) {
+        this((ClassificationRuleDAO) () -> supplier.get());
+    }
+
+    @Override
+    public String classify(NetflowDocument document) {
+        final ClassificationRequest definition = createClassificationRequest(document);
+        final String classification = classify(definition);
+        return classification;
+    }
+
+    @Override
+    public void reload() {
+        classifierList.clear();
+        ruleDAO.findAll().forEach(rule -> {
+            Classifier classifier = rule.toClassifier();
+            if (!classifierList.contains(classifier)) {
+                classifierList.add(classifier);
+            }
+        });
+        Collections.sort(classifierList, new ClassificationSorter());
+    }
+
+    @Override
+    public String classify(ClassificationRequest classificationRequest) {
+        final Optional<String> first = classifierList.stream()
+                .map(classifier -> classifier.classify(classificationRequest))
+                .filter(classifier -> classifier != null)
+                .findFirst();
+
+        // We return null instead of 'Undefined', to let the caller (e.g. rest service, or ui) decide
+        // what an unmapped definition should be named.
+        // This prevents a collision with an existing rule, which may map to 'Undefined'
+        return first.orElse(null);
+    }
+
+    protected static ClassificationRequest createClassificationRequest(NetflowDocument document) {
+        final ClassificationRequest request = new ClassificationRequest();
+        request.setProtocol(ProtocolType.createFrom(document.getIpProtocol()));
+        request.setLocation(document.getLocation());
+
+        // Decide whether to use source or dest address/port to determine application mapping
+        if (isInitiator(document)) {
+            request.setIpAddress(document.getIpv4DestAddress());
+            request.setPort(document.getDestPort());
+        } else {
+            request.setIpAddress(document.getIpv4SourceAddress());
+            request.setPort(document.getSourcePort());
+        }
+        return request;
+    }
+
+    protected static boolean isInitiator(NetflowDocument document) {
+        if (document.getSourcePort()  > document.getDestPort()) {
+            return true;
+        } else if (document.getSourcePort() == document.getDestPort()) {
+            // Tie breaker
+            return document.getIpv4SourceAddress().compareTo(document.getIpv4DestAddress()) > 0;
```

Comment:
Comparing the numeric representation of the IP address instead of their strings would make more sense here.

Reply:
Yeah, true, I forgot about that.